# This script is used to convert the matrix format generated by brocoli into the format of a sparse folder.
# weipan
# 2025-06-12
# How to use?
# python Matrix2sparse.py -i /data/workdir/panw/Data/LongBench/All_result/LR_ONT_GEX/flexiplex/BroCOLI/OUT/LB_gene_top25.txt
#                         -g /data/workdir/panw/Reference/gencode_new.v40.gtf 
#                         -m 0 -o /data/workdir/panw/LinShi/OUT


import argparse
import os
import pandas as pd
import numpy as np
import scipy.sparse as sp
from scipy.io import mmwrite
import gzip, shutil

# ==================== 函数设置 ====================
def Process_gtf(gtf_path, Mode):
    gene_isoform_dict = {}
    if gtf_path is not None:
        if 0 == Mode:
            with open(gtf_path, 'r') as f: 
                for line in f:
                    line = line.strip()
                    if not line or line.startswith('#'):
                        continue
                    fields = line.split('\t')
                    if len(fields) != 9:  
                        continue
                    feature_type = fields[2]
                    if feature_type == 'gene': 
                        gene_start = fields[3]
                        gene_end = fields[4]
                        gene_id = fields[8].split('gene_id "')[1].split('"')[0]
                        gene_name = fields[8].split('gene_name "')[1].split('"')[0]
                        chr_name = fields[0]
                        gene_isoform_dict[gene_id] = [gene_name, chr_name, gene_start, gene_end]
        else:
            with open(gtf_path, 'r') as f: 
                for line in f:
                    line = line.strip()
                    if not line or line.startswith('#'):
                        continue
                    fields = line.split('\t')
                    if len(fields) != 9:  
                        continue
                    feature_type = fields[2]   
                    if feature_type == 'transcript':
                        transcript_start = fields[3]
                        transcript_end = fields[4]
                        transcript_id = fields[8].split('transcript_id "')[1].split('"')[0]     
                        transcript_name = fields[8].split('transcript_name "')[1].split('"')[0] 
                        chr_name = fields[0]                  
                        gene_isoform_dict[transcript_id] = [transcript_name, chr_name, transcript_start, transcript_end]
    else:
        print("Please provide a gtf file!")
        exit(1)
    return gene_isoform_dict



def Generate_sparse(input_matrix_file, Mode, fileNumber, output_folder_dir, gene_isoform_dict):
    df = pd.read_csv(input_matrix_file, sep='\t', index_col=0, header=0)
    if 0 != Mode:
        df = df.drop(columns=df.columns[0])
    genes_isoforms = df.index.tolist()
    barcodes = df.columns.tolist()
    data = df.values.astype(float)
    print("----- Matrix shape:", df.shape)
    # =========== generate matrix.mtx ===========
    # find non-zero value position(0-based)
    row_ind, col_ind = np.nonzero(data)
    values = data[row_ind, col_ind]
    # build COO sparse matrix (scipy auto write 1-based to mtx)
    sparse_matrix = sp.coo_matrix((values, (row_ind, col_ind)), shape=data.shape)
    # Temporarily write to an uncompressed mtx file
    temp_mtx = os.path.join(output_folder_dir, 'matrix.mtx')
    mmwrite(temp_mtx, sparse_matrix)
    # zip .gz
    gz_mtx = temp_mtx + '.gz'
    gz_mtx = temp_mtx + '.gz'
    with open(temp_mtx, 'rb') as f_in:
        with gzip.open(gz_mtx, 'wb') as f_out:
            shutil.copyfileobj(f_in, f_out)
    os.remove(temp_mtx)
    print("INFO: matrix.mtx.gz has been generated successfully!")

    # =========== generate barcodes.tsv.gz ===========
    gz_barcodes = os.path.join(output_dir, 'barcodes.tsv.gz')
    with gzip.open(gz_barcodes, 'wt', encoding='utf-8') as f:
        for barcode in barcodes:
            f.write(barcode+'-'+str(fileNumber)+'\n')    
    print("INFO: barcodes.tsv.gz has been generated successfully!")

    # =========== generate features.tsv.gz ===========
    gz_features = os.path.join(output_dir, 'features.tsv.gz')
    with gzip.open(gz_features, 'wt', encoding='utf-8') as f:
        for gene in genes_isoforms:
            if gene in gene_isoform_dict.keys():
                gene_info = gene_isoform_dict[gene]
                f.write(f'{gene}\t{gene_info[0]}\tGene Expression\t{gene_info[1]}\t{gene_info[2]}\t{gene_info[3]}\n')
            else:
                if Mode == 0:
                    print("WARNING: Not a standard gene!")
                else:
                    f.write(f'{gene}\t{"NA"}\tGene Expression\t{"NA"}\t{"NA"}\t{"NA"}\n')
    print("INFO: features.tsv.gz has been generated successfully!")
    return 0


# ==================== 输入设置 ====================
parser = argparse.ArgumentParser(description='Convert BroCOLI matrix to seurat and scanpy!')
parser.add_argument("-i", "--input", required=True, help="input gene/isoform barcode expression matrix")
parser.add_argument("-g", "--gtf", required=True, default=None, help="gene/isoform name")
parser.add_argument("-m", "--mode", type=int, required=False, default=0, help="0-gene matrix, 1-isoform matrix")
parser.add_argument("-o", "--output", default="./filtered_feature_bc_matrix", help="output folder path")
args = parser.parse_args()

# ==================== 参数设置 ====================
input_file = args.input
input_gtf = args.gtf
output_dir = args.output
MODE = args.mode
output_dir = os.path.join(output_dir, 'filtered_feature_bc_matrix')
os.makedirs(output_dir, exist_ok=True)

input_files_list = []
if len(input_file) != 0:
    input_files_list = input_file.strip().split(',')
else:
    print("Please provide at least one input file!")
    exit(1)
print(f"----- You provide {len(input_files_list)} matrixs. -----")

if 0 == MODE:
    print("----- This is a gene matrix. -----")
else:
    print("----- This is an isoform matrix. -----")

# Process GTF;
print("----- Start to extract GTF information -----")
gene_isoform_dict = Process_gtf(input_gtf, MODE)
print("----- The GTF information extraction is successful. -----")

# Generation Files;
for file_No in range(len(input_files_list)):
    file_name = input_files_list[file_No]
    print("----- Start to process file:", file_name)
    write_No = file_No + 1
    Generate_sparse(file_name, MODE, write_No, output_dir, gene_isoform_dict)

print("Conversion completed! The file has been saved at:", output_dir)




# head -n 25 input.txt | cut -d $'\t' -f1-10 > output_25x10.txt





